usr/src/cmd/zdb/zdb.c

```diff
@@ -2301,11 +2301,18 @@ dump_label(const char *dev)
 
 		(void) snprintf(path, sizeof (path), "%s%s", ZFS_RDISK_ROOTD,
 		    dev);
-		if ((s = strrchr(dev, 's')) == NULL || !isdigit(*(s + 1)))
+		if (((s = strrchr(dev, 's')) == NULL &&
+		    (s = strchr(dev, 'p')) == NULL) ||
+		    !isdigit(*(s + 1)))
 			(void) strlcat(path, "s0", sizeof (path));
```

Comment:
I'm not entirely sure what this fixes.  It seems to me that it would prevent "s0" from being appended to fdisk partitions, which would just fail anyway.  Why not also check for ":" which is used by lofi then?